app/scripts/controllers/transactions/index.js

```diff
@@ -589,6 +589,9 @@ export default class TransactionController extends EventEmitter {
         const txMeta = cloneDeep(initialTxMeta)
         const postTxBalance = await this.query.getBalance(txMeta.txParams.from)
         txMeta.postTxBalance = postTxBalance.toString(16)
+
+        this.txStateManager.updateTx(txMeta, 'transactions#confirmTransaction - add txReceipt')
+
```

Comment:
Now `txStateManager.updateTx` will be called twice for swap transactions. Is that what we want?

Unless we want to call it twice, we should move the call to `txStateManager.updateTx` on line 586 above to an `else` clause beginning on line 596.

Reply:
Another problem is that we can't set this after an async call. This txmeta is stale, so we'd be undoing anything that occured during the async call.

To make any change after an async call you need to get another fresh copy of txmeta first, and make the change on that

Reply:
Updating it twice seems fine in principle though. Though this log message doesn't seem to accurately describe what's happening here.

Reply:
Maybe it's fine to update twice, but also unnecessary?

Reply:
I think I prefer updating twice because (after my latest changes) there is clear distinction of log messages

Reply:
> To make any change after an async call you need to get another fresh copy of txmeta first, and make the change on that

This is done in af8369f86. Also, by getting the fresh copy, I think the previous cloning is no longer necessary.

Reply:
Also, I think updating before the await is necessary to avoid the race condition that was fixed with https://github.com/MetaMask/metamask-extension/pull/9522